motr/client.h

```diff
@@ -716,6 +716,12 @@ struct m0_obj_attr {
 
 	/** Pool version fid */
 	struct m0_fid oa_pver;
+
+	/** 
+	 * Size of the object. Set this before m0_obj_init() to generate
+	 * optimal layout id during m0_entity_create().
+	 */
+	size_t        oa_obj_size;
```

Comment:
I'm not sure this is the right interface to set the optimal layout id (unit size) for an object. Did @nikitadanilov approve it?

Reply:
I am not sure I get it, did you mean to say that this is not the correct place to store the size of the object?

Reply:
@AbhishekSahaSeagate, Motr does not store object sizes. As far as I understand from the comment above the `oa_obj_size`, it is intended as an user interface to set the optimal unit size for the object. In other words, user have to set this attribute so that the optimal unit size is calculated automatically. But it does not mean that the actual object size will correspond to this attribute (it might be greater or less, object may be appended later, for example, and this parameter will never be updated).

In my opinion, it is very confusing interface to set the optimal unit size.

Reply:
Yes, I agree that the `oa_obj_size` attribute may become stale when the object is appended or truncated but the intention of this field is to derive an optimal layout id during object creation and not to be used again considering S3 won't append/truncate the object after creating it.

Would converting this into an API and then letting the user set it during m0_obj_init() would be a better approach?
@andriytk , @sg-shankar

Reply:
Renaming it to `oa_buf_size` might improve the situation a little. User can specify which buffer size will be used to do the I/O on the object from which the optimal unit size can be calculated. But still, this is not really an attribute of the object itself. Another client from another machine can do the I/O on the object and use completely different buffer size.

Layout id can be set explicitly by user via `m0_obj_init()` API. So we better add some new function which can calculate optimal layout id for user based on object size and pool id. See an example of such interface here - https://github.com/Seagate/cortx-motr/blob/ae143131c275a3a30560386a9e4180817f52bd44/bindings/go/mio/mio.go#L333-L343

We can also add another function which would calculate the optimal block size for the I/O operation for user based on the buffer size user affords to use in his application, see an example here:
https://github.com/Seagate/cortx-motr/blob/ae143131c275a3a30560386a9e4180817f52bd44/bindings/go/mio/mio.go#L367-L368

Reply:
This size is just a hint. It is not mandatory, or fixed. S3 server knows the final object size when it creates an object.
So, this size can be used as a hint to get an optimal layout.

Reply:
I do understand this is a hint. Still, it's rather an awkward and confusing way of specifying a hint. Because, as I mentioned above, this is not an object size (but I/O op block size). Neither it is an object attribute.

Reply:
Yes, we had good discussion with S3 team, @nikitadanilov on this API change and what's the best way to do this. During discussion this was agreed upon change to the API. 

This discussion was captured in following confluence [document](https://seagate-systems.atlassian.net/wiki/spaces/PUB/pages/103055415/Fault+Tolerance+Handling+Analysis): ref 4.b and 4.c